include/experimental/__p2630_bits/submdspan_mapping.hpp

```diff
@@ -0,0 +1,199 @@
+MDSPAN_INLINE_FUNCTION
+constexpr auto
+construct_sub_strides(const SrcMapping &src_mapping,
+                      index_sequence<InvMapIdxs...>,
+                      const tuple<slice_strides...> &slices_stride_factor) {
+  using index_type = typename SrcMapping::index_type;
+  return array<typename SrcMapping::index_type, sizeof...(InvMapIdxs)>{
+      (static_cast<index_type>(src_mapping.stride(InvMapIdxs)) *
+       static_cast<index_type>(get<InvMapIdxs>(slices_stride_factor)))...};
+}
+} // namespace detail
+
+//**********************************
+// layout_left submdspan_mapping
+//*********************************
+namespace detail {
+
+// Figure out whether to preserve layout_left
+template <class IndexSequence, size_t SubRank, class... SliceSpecifiers>
+struct preserve_layout_left_mapping;
+
+template <class... SliceSpecifiers, size_t... Idx, size_t SubRank>
+struct preserve_layout_left_mapping<index_sequence<Idx...>, SubRank,
+                                    SliceSpecifiers...> {
+  constexpr static bool value =
+      // Preserve layout for rank 0
+      (SubRank == 0) ||
+      (
+          // Slice specifiers up to subrank need to be full_extent_t - except
+          // for the last one which could also be tuple but not a strided index
+          // range slice specifiers after subrank are integrals
+          ((Idx > SubRank - 1) || // these are only integral slice specifiers
+           (is_same_v<SliceSpecifiers, full_extent_t>) ||
+           ((Idx == SubRank - 1) &&
+            is_convertible_v<SliceSpecifiers, tuple<size_t, size_t>>)) &&
+          ...);
+};
+} // namespace detail
+
+// Actual submdspan mapping call
+template <class Extents, class... SliceSpecifiers>
+MDSPAN_INLINE_FUNCTION
+constexpr auto
+submdspan_mapping(const layout_left::mapping<Extents> &src_mapping,
+                  SliceSpecifiers... slices) {
+
+  // compute sub extents
+  using src_ext_t = Extents;
+  auto dst_ext = submdspan_extents(src_mapping.extents(), slices...);
+  using dst_ext_t = decltype(dst_ext);
+
+  // figure out sub layout type
+  constexpr bool preserve_layout = detail::preserve_layout_left_mapping<
+      decltype(make_index_sequence<src_ext_t::rank()>()), dst_ext_t::rank(),
+      SliceSpecifiers...>::value;
+  using dst_layout_t =
+      conditional_t<preserve_layout, layout_left, layout_stride>;
+  using dst_mapping_t = typename dst_layout_t::template mapping<dst_ext_t>;
+
+  if constexpr (is_same_v<dst_layout_t, layout_left>) {
+    // layout_left case
+    return mapping_offset<dst_mapping_t>{
+        dst_mapping_t(dst_ext),
+        static_cast<size_t>(src_mapping(detail::first_of(slices)...))};
+  } else {
+    // layout_stride case
+    #if MDSPAN_HAS_CXX_20
+    using inv_map_t = typename detail::inv_map_rank<0, index_sequence<>,
+                                                    SliceSpecifiers...>::type;
+    #else
+    using inv_map_t = decltype(detail::inv_map_rank(
```

Comment:
While the function version of `inv_map_rank` has a shorter and simpler implementation than the struct version, not everyone is a fan of functions whose only purpose is to deduce a type ("`decltype` everywhere").  Christian points out offline that we also don't know which approach would compile faster.  Thus, I think we should leave both of these paths in the code, at least until we have compilation benchmark results that suggest one over the other.

Reply:
Counterargument: We don't actually need `inv_map_t` here, we need an object of that type.  In that case, why not just call the function to get an `inv_map` object?

```c++
auto inv_map = detail::inv_map_rank(
  integral_constant<size_t,0>{},
  index_sequence<>{},
  (declval<SliceSpecifiers>())...);
return mapping_offset<dst_mapping_t>{
  dst_mapping_t(
    dst_ext,
    detail::construct_sub_strides(
      src_mapping,
      inv_map,
      tuple{detail::stride_of(slices)...})),
  static_cast<size_t>(src_mapping(detail::first_of(slices)...))};
```